src/librustc_codegen_ssa/mir/mod.rs

```diff
@@ -636,6 +636,44 @@ fn arg_local_refs<'a, 'tcx: 'a, Bx: BuilderMethods<'a, 'tcx>>(
                     DUMMY_SP
                 );
             }
+
+            if let Some(generator_layout) = &mir.generator_layout {
+                let (def_id, gen_substs) = match closure_layout.ty.sty {
+                    ty::Generator(def_id, substs, _) => (def_id, substs),
+                    _ => bug!("generator layout without generator substs"),
+                };
+                let state_tys = gen_substs.state_tys(def_id, tcx);
+
+                let upvar_count = mir.upvar_decls.len();
+                for (i, (decl, ty)) in generator_layout.fields.iter().zip(state_tys).enumerate() {
+                    let name = if let Some(name) = decl.name { name } else { continue };
+                    let byte_offset_of_var_in_env = closure_layout.fields
+                        .offset(upvar_count + i + 1)
+                        .bytes();
+
+                    let ops = bx.debuginfo_upvar_decls_ops_sequence(byte_offset_of_var_in_env);
+
+                    // The environment capture can be indirect.
+                    let ops = if env_ref { &ops[..] } else { &ops[1..] };
+
+                    // The variable itself is never indirect
+                    let ops = &ops[..ops.len() - 1];
+
+                    let variable_access = VariableAccess::IndirectVariable {
+                        alloca: place.llval,
+                        address_operations: &ops
+                    };
+                    bx.declare_local(
+                        &fx.debug_context,
+                        name,
+                        ty,
+                        scope,
+                        variable_access,
+                        VariableKind::LocalVariable,
+                        DUMMY_SP
+                    );
+                }
+            }
```

Comment:
Probably should factor this out to a method instead of duplicating it

Reply:
Updated to first combine the `UpvarDecl`s and `LocalDecl`s into a single iterable that then has the locals generated for it.